Feat/fat decoding

NyuBlara opened the pull request:
Implemented FAT files decoding.

Tested on Professor Layton 1 through 3, and Zelda Phantom Hourglass with no problems. 

Code may break certain coding style guidelines, I'm ready to adapt to them before merge.

rblard committed:
First step in implementing FAT file decoding : brute-force extraction without FNT lookup.

The FAT alone contains enough information to extract every file in a
folder chosen by the user. However, reference to the file name table
(FNT) is needed to restore the files' names and the directory structure
(and avoid dumping useless data that may be left over in the data
section).

This commit only implements the dumping of all data by incrementing
through the FAT and writing every section marked by a FAT range into a
separate file, giving them incremental names.

The next commit will use the FNT to name them properly.

NyuBlara committed:
Proper FAT decoding : FNT lookup and directory structure

Code comments explain the algorithm (which was understood thanks to the 
wonderfully simple "FNT-Tool" script at : 
https://github.com/RoadrunnerWMC/FNT-Tool) ; the implementation is more 
or less a C++ port of the Python code.

Everything may still be a little dirty, but it works !

NyuBlara committed:
Updated README!

NyuBlara committed:
Updated README!

Luca1991 commented:
Thank you very much for your contribution, I was watching your progress yesterday :P

I'll make sure to have a look at your code ASAP and merge this into master (or let you know if there is anything to fix).

NyuBlara commented:
You're welcome ! Very glad to do it. Also very glad to see you review the PR, I was a bit scared because the last commits were so long ago haha. 

I'll be on the lookout for your comments !

NyuBlara committed:
MINOR compliance modifications 

- Removed extra entry in .gitignore
- Removed superfluous debug inclusions 
- Replaced magic values by preprocessor macros

Luca1991 commented:
Perfect, thanks to your compliance fixes, we can now merge your work! Thank you :)
